### PR TITLE
Stipend module implementation

### PR DESCRIPTION
**Pull request by OlivierBBB**
WIP, implements the following
- #98

### CODE REVIEW

**Review by letypequividelespoubelles — APPROVED** [review]

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+    }
+  }
+
+  private StpChunk getCreateData(final MessageFrame frame) {
+    final Address to = getDeploymentAddress(frame);
+    final Long gasRemaining = frame.getRemainingGas();
+    final Long gasMxp = getGasMxpCreate(frame);
+    final Long gasPrelim = GasConstants.G_CREATE.cost() + gasMxp;
+    return new StpChunk(
+        OpCode.of(frame.getCurrentOperation().getOpcode()),
```
Comment (delehef) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java:117]: `this.hub.opCode()`

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+        gasPrelim,
+        gasRemaining < gasPrelim,
+        gasMxp,
+        frame.getWorldUpdater().get(frame.getContractAddress()).getBalance(),
+        to,
+        Bytes32.leftPad(frame.getStackItem(0)));
+  }
+
+  private StpChunk getCallData(final MessageFrame frame) {
+    final OpCode opcode = OpCode.of(frame.getCurrentOperation().getOpcode());
```
Comment (delehef) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java:128]: `this.hub.opCode()`

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+        gasMxp,
+        frame.getWorldUpdater().get(frame.getContractAddress()).getBalance(),
+        to,
+        value,
+        toExists,
+        toWarm,
+        Bytes32.leftPad(frame.getStackItem(0)));
+  }
+
+  private boolean cctv(OpCode opCode) {
```
Comment (delehef) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java:166]: cctv = « coucou tu veux voir [ma bite] » ? “China Central TeleVision”?

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+    long gasMxp = 0;
+    long offset = Words.clampedToLong(frame.getStackItem(1));
+    long length = Words.clampedToLong(frame.getStackItem(2));
+    long currentMemorySizeInWords = frame.memoryWordSize();
+    long updatedMemorySizeInWords = frame.calculateMemoryExpansion(offset, length);
+    if (currentMemorySizeInWords < updatedMemorySizeInWords) {
+      // computing the "linear" portion of CREATE2 memory expansion cost
+      long G_mem = GasConstants.G_MEMORY.cost();
+      long squareCurrent = (currentMemorySizeInWords * currentMemorySizeInWords) >> 9;
+      long squareUpdated = (updatedMemorySizeInWords * updatedMemorySizeInWords) >> 9;
```
Comment (delehef) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java:181]: All these can be `final`

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+      gasMxp +=
+          G_mem * (updatedMemorySizeInWords - currentMemorySizeInWords)
+              + (squareUpdated - squareCurrent);
+    }
+    if (OpCode.of(frame.getCurrentOperation().getOpcode()) == OpCode.CREATE2) {
+      long lengthInWords = (length + 31) >> 5; // ⌈ length / 32 ⌉
```
Comment (delehef) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java:187]: Same

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+    int offset = cctv(OpCode.of(frame.getCurrentOperation().getOpcode())) ? 1 : 0;
+    long cdo = Words.clampedToLong(frame.getStackItem(2 + offset)); // call data offset
+    long cds = Words.clampedToLong(frame.getStackItem(3 + offset)); // call data size
+    long rdo = Words.clampedToLong(frame.getStackItem(4 + offset)); // return data offset
+    long rdl = Words.clampedToLong(frame.getStackItem(5 + offset)); // return data size
+
+    long memSize = frame.memoryWordSize();
+    long memSizeCallData = frame.calculateMemoryExpansion(cdo, cds);
+    long memSizeReturnData = frame.calculateMemoryExpansion(rdo, rdl);
+    long maybeNewMemSize = max(memSizeReturnData, memSizeCallData);
```
Comment (delehef) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java:206]: Same

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+
+    if (memSize < maybeNewMemSize) {
+      // computing the "linear" portion of CREATE2 memory expansion cost
+      long G_mem = GasConstants.G_MEMORY.cost();
+      long squareCurrent = (memSize * memSize) >> 9;
+      long squareUpdated = (maybeNewMemSize * maybeNewMemSize) >> 9;
```
Comment (delehef) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java:212]: Same

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/StpTrace.java
Comment (delehef) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/StpTrace.java:1]: This file should not exist

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+    this.chunks.pop();
+  }
+
+  @Override
+  public int lineCount() {
+    int traceRowSize = 0;
+    for (StpChunk chunk : this.chunks) {
+      traceRowSize += ctMax(chunk) + 1;
+    }
+    return traceRowSize;
```
Comment (delehef) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java:73]: `return this.chunks.stream().mapToInt(chunk -> ctMax(chunk) + 1).sum();`

File: arithmetization/src/main/java/net/consensys/linea/zktracer/types/OpCodeUtils.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package net.consensys.linea.zktracer.types;
+
+import net.consensys.linea.zktracer.opcode.OpCode;
+
+public class OpCodeUtils {
+  public static boolean isCreate(OpCode opCode) {
```
Comment (delehef) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/types/OpCodeUtils.java:21]: Move that in `OpCode.java`

**Review by delehef — CHANGES_REQUESTED** [review]

File: arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java (hunk 3 above)
Comment (letypequividelespoubelles) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/module/stp/Stp.java:166]: call can transfer value, directly from the spec. It goes with @OlivierBBB's SUX.

**Review by delehef — APPROVED** [review]

**Pull request merged**